tools/gn.rs

```diff
@@ -136,3 +142,105 @@ fn normalize_path<T: AsRef<Path>>(path: T) -> String {
     .map(|c| if path::is_separator(c) { '/' } else { c })
     .collect()
 }
+
+fn maybe_print_rerun_if_changed(dep: &str, gn_out_dir: &PathBuf) {
+  let d = gn_out_dir.join(dep);
+  println!(
+    "cargo:rerun-if-changed={}",
+    d.into_os_string().into_string().unwrap()
+  );
+}
+
+fn gn_command(root: &PathBuf) -> Command {
+  // Use DENO_GN_PATH if present.
+  let p = env::var_os("DENO_GN_PATH").unwrap_or_else(|| {
+    if cfg!(target_os = "windows") {
+      root.join("third_party/depot_tools/gn.bat")
+    } else {
+      root.join("third_party/depot_tools/gn")
+    }.into_os_string()
+  });
+  Command::new(p)
+}
+
+fn ninja_command(root: &PathBuf) -> Command {
+  // Use DENO_NINJA_PATH if present.
+  let p = env::var_os("DENO_NINJA_PATH").unwrap_or_else(|| {
+    root.join("third_party/depot_tools/ninja").into_os_string()
+  });
+  Command::new(p)
+}
+
+fn rerun_if_changed(gn_target: &str, gn_out_dir: &PathBuf, root: &PathBuf) {
+  // Notes on cargo:rerun-if-changed. The feature is very sensitive. If you do
+  // not provide the path in the correct form, the build will always rebuild.
+  // - Any invalid filename will cause rebuild.
+  // - Do not put the path in quotes.
+  // - Relative paths work when relative to the Cargo.toml directory.
+  // - Absolute paths work (no quotes)
+
+  // Example: gn desc target/debug/ cli:deno_deps runtime_deps
+  let mut gn = gn_command(root);
+  let output = command_env(&mut gn, root)
+    .arg("desc")
+    .arg(gn_out_dir)
+    .arg(gn_target)
+    .arg("runtime_deps")
+    .output()
+    .expect("gn desc failed");
+  let stdout = std::str::from_utf8(&output.stdout).unwrap();
+  for dep in stdout.lines() {
+    maybe_print_rerun_if_changed(dep, gn_out_dir);
+  }
+
+  // Example: ninja -C target/debug/ cli:deno_deps -t deps
```

Comment:
Note that the target is ignored by ninja. This just prints the deps database for all targets that ninja knows about.